src/trace/statsd.zig

```diff
@@ -0,0 +1,463 @@
+
+        self.* = undefined;
+    }
+
+    pub fn emit(
+        self: *StatsD,
+        events_metric: []const ?EventMetricAggregate,
+        events_timing: []const ?EventTimingAggregate,
+    ) void {
+        assert(events_metric.len == self.events_metric_emitted.len);
+
+        // This really should not happen; it means we're emitting so many packets, on a short
+        // enough emit timeout, that the kernel hasn't been able to process them all (UDP doesn't
+        // block or provide back-pressure like a TCP socket).
+        //
+        // Keep it as a log, rather than assert, to avoid the common pitfall of metrics killing
+        // the whole system.
+        if (self.buffer_completions.executing() != 0) {
+            log.err("{} / {} packets still in flight; trying to continue", .{
+                self.buffer_completions.executing(),
+                packet_count_max,
+            });
+        }
+
+        if (self.implementation == .udp and self.implementation.udp.send_callback_error_count > 0) {
+            log.warn(
+                "failed to send {} packets",
+                .{self.implementation.udp.send_callback_error_count},
+            );
+            self.implementation.udp.send_callback_error_count = 0;
+        }
+
+        var buffer_completion = self.buffer_completions.acquire() orelse {
+            log.err("insufficient packets to emit any metrics", .{});
+            assert(self.implementation == .udp);
+            return;
+        };
+
+        var buffer_written: usize = 0;
+        inline for (.{ events_metric, events_timing }) |events| {
+            for (events, 0..) |event_new_maybe, i| {
+                const event_new = event_new_maybe orelse continue;
+                if (@TypeOf(event_new) == EventMetricAggregate) {
+                    const event_old = &self.events_metric_emitted[i];
+                    if (event_old.* != null and event_old.*.?.value == event_new.value) continue;
+                    event_old.* = event_new;
+                } else {
+                    comptime assert(@TypeOf(event_new) == EventTimingAggregate);
+                }
+
+                const stats = switch (@TypeOf(event_new)) {
+                    EventMetricAggregate => [_]Stat{.{ .metric = .{ .aggregate = event_new } }},
+                    EventTimingAggregate => [_]Stat{
+                        .{ .timing = .{ .aggregate = event_new, .stat = .min } },
+                        .{ .timing = .{ .aggregate = event_new, .stat = .max } },
+                        .{ .timing = .{ .aggregate = event_new, .stat = .avg } },
+                        .{ .timing = .{ .aggregate = event_new, .stat = .sum } },
+                        .{ .timing = .{ .aggregate = event_new, .stat = .count } },
+                    },
+                    else => unreachable,
+                };
+
+                for (stats) |stat| {
+                    buffer_written += format_metric(
```

Comment:
Ok, yeah, this feels much simpler now, but I got another crazy idea! 

What if we don't have a per-write buffer, but have just one huge, contiguous buffer for _all_ metrics?

Then, we just format _all_ metrics into this single buffer without any retries, and:


* Either maintain a bounded array of offsets of boundaries, to do one vectored write at the end
* Or maintain the running offset sent, and send a slice as soon as we have an mtg-worth of data

The drawback would be that we'd have to wait for _all_ packets to come through before starting the next loop, but that seems fine, as we _do_ expect all the packets to go through.